### ISSUE TITLE
Solve user agent camera/microphone double-mute

### ISSUE THREAD

**Issue by jan-ivar** [w3c/mediacapture-extensions#39]
User agent mute-toggles for camera & mic can be useful, yielding enhanced privacy (no need to trust site), and quick access (a sneeze coming on, or a family member walking into frame?)
- Safari has pause/resume in its URL bar
- Firefox has global cam/mic mute toggles behind a pref (set `privacy.webrtc.globalMuteToggles` in about:config)
- Chrome has recently opened [an issue](https://crbug.com/1263133) discussing it.

It's behind a pref in Firefox because:
 1. The **_double-mute problem_**: site-mute + ua-mute = 4 states, where 3 produce no sound (_"Can you hear me now?"_)
 2. UA-mute of microphone interferes with _"Are you talking?"_ features
 3. Some sites (Meet) _stop_ camera to work around [crbug 642785](https://crbug.com/642785) in Chrome, so there's no video track to UA-mute

This image is titled: "Am I muted?"
<img width="700" src="https://user-images.githubusercontent.com/3136226/139155537-88fec96f-241a-4107-af68-c65da91570fc.png">

### This issue is only about (1) the double-mute problem.

We determined we can only solve the double-mute problem by involving the site, which requires standardization.

The idea is:
 1. If the UA mutes or unmutes, the site should update its button to match.
 2. If the user unmutes using the site's button, the UA should unmute(!)

The first point requires no spec change: sites can listen to the [mute](https://w3c.github.io/mediacapture-main/#event-mediastreamtrack-mute) and [unmute](https://w3c.github.io/mediacapture-main/#event-mediastreamtrack-unmute) events on the track (but they don't).

The second point is key: if the user sees the site's button turn to "muted", they'll expect to be able to click it to unmute.

This is where it gets tricky, because we don't want to allow sites to unmute themselves at will, as this defeats any privacy benefits.
 
The proposal here is:
```idl
partial interface MediaStreamTrack {
  undefined unmute();
}
```
It would throw `InvalidStateError` unless it has transient activation, is fully active, and has focus. User agents may also throw `NotAllowedError` for any reason, but if they don't then they must unmute the track (which will fire the unmute event).

This should let user agents that wish to develop UX without the double-mute problem.

**Comment by jan-ivar** [comment]
That's a good idea. Having the method be asynchronous would allow this.
```idl
partial interface MediaStreamTrack {
  Promise<undefined> unmute();
```
The goal of a spec here is to allow innovation in this space, without arriving at specific UX. It could be a prompt, or maybe a toast message is enough.

I think a lot of users would be surprised to learn that when they mute microphone or camera on a web site today, they have zero assurances that it actually happens. Well-behaved websites have lulled most users into feeling they're in charge when they're not. Most probably don't consider that the website may turn camera or microphone back on at any time as long as the page is open.

The page doesn't even need to be open anymore: a different (origin) accomplice page can reopen/navigate to it without user interaction at a later point, since gUM doesn't require transient activation.

We dropped the ball on transient activation in gUM. Having better UA-integrated muting with transient activation might give us a second chance to rectify some privacy concerns.

For instance, a UA could choose to mute for the user if it detects a page accessing camera or microphone on pageload or without interaction.

**Comment by youennf** [comment]
> ```idl
> Promise<undefined> unmute();
> ```

`unmute()` method seems fine.
I wonder whether we should not try to introduce a `mute()` method as well.
This would allow the website to mute itself and have its UI synchronised with OS/UA UI.

Also, in Safari, the mute/unmute is page wide, which means that all microphone tracks are either muted or unmuted.
This does not align particularly well with unmute being at the track level.
Maybe introducing unmute/mute at navigator.mediaDevices level would be good enough?

**Comment by jan-ivar** [comment]
> I wonder whether we should not try to introduce a mute() method as well.

I fear this would create confusion and a false symmetry suggesting muting is under application control when it is not.

> This would allow the website to mute itself and have its UI synchronised with OS/UA UI.

Applications have `track.enabled` for this. UAs _"[MAY](https://w3c.github.io/mediacapture-main/#life-cycle)"_ turn off indicators when this _"brings all tracks connected to the device to be ... disabled,"_ which [Firefox has done since 2018](https://blog.mozilla.org/webrtc/better-privacy-on-camera-mute-in-firefox-60/). This is [crbug 642785](http://crbug.com/642785), I dunno if webkit has one.

Here's a [fiddle for toggling multiple tracks](https://jsfiddle.net/jib1/pmgst13v/) demonstrating Firefox updating its camera URL bar indicator and OS/hardware light (modulo [bug 1694304](https://bugzil.la/1694304) on Windows for the hardware light), whenever the number of enabled tracks go to zero.

I suggest leaving mute to UAs and concentrating on how apps can signal interest to unmute, to solve the issue at hand. 

> Also, in Safari, the mute/unmute is page wide,

Do you mean page wide or document wide? What about iframes?

> ... which means that all microphone tracks are either muted or unmuted. This does not align particularly well with unmute being at the track level.

Maybe we could use constraints? 🙂

> Maybe introducing unmute/mute at navigator.mediaDevices level would be good enough?

Maybe, except `muted` is a property of the track, not `mediaDevices`.

**Comment by jan-ivar** [comment]
There's also the issue of multiple cameras. If we end up with `navigator.mediaDevices.unmute(track)` that would really suck.

**Comment by jan-ivar** [comment]
> track.unmute() makes sense if we think this is useful for all source types.

It was a WG design choice to reuse MST for other sources to avoid inheritance. The cost of that is living with the fact that not all sources have all track abilities, NOT that tracks only have abilities shared by all sources.

getUserMedia returns camera and microphone tracks, so adding attributes, methods and constraints specific to camera and microphone should be fine. If it's not, then time to split inheritance.

E.g. track.[muted](https://w3c.github.io/mediacapture-main/getusermedia.html#track-muted) only makes sense for camera and microphone, and `track.unmute()` fits with that.

Other sources do not have the double-mute problem, so to not complicate discussion, let's not discuss them here.

> This is a MAY though.

[Here's](https://w3c.github.io/mediacapture-main/#life-cycle) a SHOULD: _"When a "[live](https://w3c.github.io/mediacapture-main/#idl-def-MediaStreamTrackState.live)", [unmuted](https://w3c.github.io/mediacapture-main/#track-muted), and [enabled](https://w3c.github.io/mediacapture-main/#track-enabled) track sourced by a device exposed by [getUserMedia()](https://w3c.github.io/mediacapture-main/#dom-mediadevices-getusermedia) becomes either [muted](https://w3c.github.io/mediacapture-main/#track-muted) or [disabled](https://w3c.github.io/mediacapture-main/#track-enabled), and this brings all tracks connected to the device (across all [navigables](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable) the user agent operates) to be either muted, disabled, or stopped, then the UA SHOULD relinquish the device within 3 seconds..."_

> Setting all tracks of the same source with enabled = false does not mean track.muted will switch to true, this is left to UA which does not seem great for interop.

Why would a UA mute an in-view application just because it disabled its tracks? That would be terrible for web compat.

To clarify, Mozilla needs no spec changes to solve turning off privacy indicators or camera light. Our view is the path to interop there is changing the MAY and SHOULD to MUST. But please let's discuss that in a separate issue.

> In Firefox mode, I would guess that muted would be set to true later on, when the application sets enabled = true on one of the track, in which case the application will then have to call unmute. This is not simple.

No that is not our plan. As explained in the OP we have a `privacy.webrtc.globalMuteToggles` pref in about:config which turns on the global user-facing mute controls shown. <img height="18" alt="image" src="https://user-images.githubusercontent.com/3136226/233522632-09c8dcc8-bf94-4fc4-be23-ac028b4d7eb3.png">, and we want to arm sites with tools to unmute themselves better to prepare for UA features like this.

Sorry for any misunderstanding, but it's not my intent to standardize UA muting here, only application-induced unmuting. Muting remains up to user agents, and I think it is important for privacy that they be allowed to continue to own that problem.

The scope of the proposal in the OP (and this issue) was to arm applications with tools to unmute themselves _IF_ the user agent mutes them, not define _when_ user agents mute them.

**Comment by jan-ivar** [comment]
> `unmute` already introduces this potential confusion about who has control, hence the `requestMute/requestUnmute` name.

We have `getUserMedia`, not `requestUserMedia`, and that doesn't seem to confuse anybody.

`NotAllowedError` seems clear about who has control.

**Comment by guidou** [comment]
I'd like to revive this discussion, since these types of system-level controls (either at the UA or the OS) are becoming more common and we have observed that they create a lot of confusion for users.
Like @jan-ivar says, there are two issues here:
> If the UA mutes or unmutes, the site should update its button to match.
> If the user unmutes using the site's button, the UA should unmute(!)

However, I disagree on the statement that `The first point requires no spec change: sites can listen to the [mute](https://w3c.github.io/mediacapture-main/#event-mediastreamtrack-mute) and [unmute](https://w3c.github.io/mediacapture-main/#event-mediastreamtrack-unmute) events on the track (but they don't)` 

The reason sites don't listen to the `mute` and `unmute` event is that mute and unmute can be triggered by other causes and if the application cannot know that those events (and the`muted` attribute) are caused by the double-mute problem, it cannot react appropriately. The spec says muted means _live samples are not made available to the MediaStreamTrack_, which is not specific to UA/OS-level mute controls. In Chrome specifically, muted means a track is not getting frames for any reason (and system-level mute has never been one of those reasons in practice).

This has become a major problem for VC applications and I think we need to solve it properly.
I think we can iterate on several of the proposals made in this thread, which look like very promising IMO.

cc @eladalon1983

**Comment by eladalon1983** [comment]
The problem as I see it is that users can mute through multiple sources - app, UA, OS and hardware. The propagation of state through these layers is presently incomplete - an opportunity for us to earn our keep.

In the high-level, I think we have to provide two mechanisms:
1. Sites listen for mute-status changes from upstream sources (UA, OS and hardware, in that order).
2. Sites control mute-status in upstream sources.

### 1. Listen
The principle here should be fairly uncontroversial.
For the concrete mechanism, I agree with Guido that mute events are not currently well-suited. Either of the following would work for me:
1. Bespoke events.
2. Revive the [idea](https://github.com/w3c/mediacapture-region/issues/9#issuecomment-1022334881) of a `MuteCause`/`MuteReason`, so that the same mechanism would service both this issue as well as similar ones (see link).

I prefer no2. To start the ball rolling on a concerete proposal:
```webidl
enum MuteCause {
  "unspecified",  // Catch-all default.
  "operating-system-choice",
  "user-agent-choice",
  // Extensible to hardware, and to -issue if it's not a -choice.
};

interface MuteEvent : Event {
  /* Exercise for the reader */
};

partial interface MediaStreamTrack {
  // Note that multiple causes might apply concurrently.
  readonly attribute sequence<MuteCause> causes;
};
```

### 2. Control
There's some potential for controversy here, but I think we can resolve it.

Jan-Ivar proposed:
> If the user unmutes using the site's button, the UA should unmute(!)

While I'm sure VC apps would be delighted to have such control, I am afraid that no security/privacy department in any user agent would ever approve it (unless we add some UA-prompt; foreshadowing). Jan-Ivar [suggested](https://github.com/w3c/mediacapture-extensions/issues/39#issue-1037935336) transient activation and focus as necessary gating mechanisms. These are fine requirements, but they are not sufficient as any transient activation would look identical to the user agent here, possibly subverting the user's actual intentions if they clicked on a mislabelled button. I'd suggest requiring also a [PEPC](https://github.com/WICG/PEPC/blob/main/explainer.md)-like prompt. Reasonable app code would then look something like this:
```js
unmuteButton.addEventListener('click', "unmuteClicked();");

async function unmuteClicked() {
  // If necessary, prompt the user to unmute at UA-level etc.
  if (upstreamUnmute) {
    try {
      await track.unmute();
    } catch (error) {
      return;
    }
  }

  // Proceed with the "normal" unmuting in the app.
  // * Resume remote transmission of media.
  // * Change UX to reflect that clicking the button now means "mute".
  // * Update internal state.
}
```

**Comment by youennf** [comment]
I am not sure how much we need a mute reason. Distinct requestUnmute failures might be sufficient.

**Comment by eladalon1983** [comment]
1. You need a MuteReason because mute can happen for inactionable reasons too, like the source not having any new frames to deliver.
2. Applications might not wish requestUnmute() themselves, but rather provide some reminder/hint to the user about where they muted from (either browser or operating system), and leave it to the user to unmute if they wish.

**Comment by youennf** [comment]
The mute reasons may vary within the muted period and firing mute events when only the reason is changing is not really appealing.
The flow of user trying to unmute and app providing the hint based on the failure seems to cover most cases (and it does not preclude adding reasons in the future if we discover this is useful).

**Comment by guidou** [comment]
> The mute reasons may vary within the muted period and firing mute events when only the reason is changing is not really appealing.

I'm OK with the mute event only firing when the muted attribute changes (not the muted reason attribute). WDYT?
The main point iis that the muted attribute in its current form is not enough to solve this problem.
Having a reason looks like a good way to make the muted attribute useful to solve this problem. Otherwise, we need a new attribute or a different API surface. 

> The flow of user trying to unmute and app providing the hint based on the failure seems to cover most cases (and it does not preclude adding reasons in the future if we discover this is useful).

Does this mean you support the approach of having an attribute for the mute cause?

**Comment by eladalon1983** [comment]
> The mute reasons may vary within the muted period and firing mute events when only the reason is changing is not really appealing.

Why is it not appealing to fire a mute event when the set of reasons changes? (Note that we have a separate unmute event already, btw.)

**Comment by youennf** [comment]
> Does this mean you support the approach of having an attribute for the mute cause?

I see this as a potential improvement while I see an API to request unmute as a blocker.
I would focus on the unblocking API.

I also think that an API to request capture to be muted would be useful. The current approach (use enabled=false on all tracks of a device), is a bit tedious and might face backward compatibility issues.

**Comment by guidou** [comment]
> I see this as a potential improvement while I see an API to request unmute as a blocker.
I would focus on the unblocking API.

I agree. Let's focus on that first.

> I also think that an API to request capture to be muted would be useful. The current approach (use enabled=false on all tracks of a device), is a bit tedious and might face backward compatibility issues.

Also agree. This requires some more thinking because system-level muting is not necessarily equivalent to muting a source or a set of tracks.

**Comment by eladalon1983** [comment]
I'm having some trouble parsing the last few messages on this thread. If we're all in agreement that we want to add an API exposing the OS-mute-state, then I'll gladly present something in the next available interim opportunity. Specifically, I'd like to present my proposal [here](https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-1805921604). Before I do - @youennf, you have [said](https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-1806094005) that firing an event whenever the reason changes is unappealing. I'd [still](https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-1807725727) like to understand why; understanding would help drive a better presentation in said interim.

**Comment by youennf** [comment]
@guidou and I seem to agree on focusing on the following items (sorted by priority):
- Expose something like `requestUnmute` as this is required to solve double-mute.
- Evaluate the need for `requestMute`.
- Evaluate the need for a mute reason.

Getting back to `requestUnmute`, here are some possible API shapes (all promise based):
- `track.requestUnmute()` // all tracks of the related device
- `navigator.mediaDevices.requestUnmute(track)` // all tracks of the related device
- `inputDeviceInfo.requestUnmute()` // does not work for screen capture so probably not ok.
- `navigator.mediaDevices.requestUnmute(captureKind)` // all devices of the given kind.

I would tend to unmute at the device level.

**Comment by jan-ivar** [comment]
The OP assumes users can always unmute. If there's a second actor controlling mute that the user agent cannot affect, then double-mute likely remains the best way to handle that piece. Otherwise we get:

<img width="309" alt="image" src="https://github.com/w3c/mediacapture-extensions/assets/3136226/88048e02-24b7-4e82-a185-18d30d35cc0a"><br>
A. Unmuted B. Muted (actionable) C. Muted (unactionable)

Today's apps show C as A.¹ To maintain this, they'd need to distinguish `"actionable"` from `"unactionable"` mute.

I'd support those two `MuteReason`s, but would avoid exposing more cross-origin correlatable information than that. I don't mind re-firing mute when reason changes.

Regarding method shape, I think `track.unmute()` is all it takes, because
1. every method that can fail with `NotAllowedError` is a request, and
2. [muted](https://w3c.github.io/mediacapture-main/#dom-mediastreamtrack-muted) is already a non-locally-configurable property of the track, so `umute()` would be a non-locally-contained action, a signal to the UA, which in turn ultimately controls the scope of that action (though it might be useful to define a minimally affected scope). How many documents a UA ultimately enforces mute or unmute upon seems implementation-defined.

---
<sub>1. A case could be made for showing C as B, but then nothing happens when the user clicks on it, which seems undesirable. This is an app decision of course.</sub>

**Comment by eladalon1983** [comment]
> @guidou and I seem to agree on focusing on the following items (sorted by priority):

We would have to ask Guido to see if you two agree, but I personally disagree with your prioritization. I think it's necessary for Web applications like Meet to know when the input devices is muted through the upstream (browser or OS), or else they can't update their UX to show the input device is muted, which means the user won't press it, which means they can't call `requestUnmute()` - whatever its shape. The very first step is for the application to know the upstream is muted. That's the top priority.

**Comment by eladalon1983** [comment]
<img width="948" alt="image" src="https://github.com/w3c/mediacapture-extensions/assets/22117736/94bb0678-b5c9-420a-a1d1-af065e61f164">

Above - an illustration. Without a `MuteCause` or a similar API, how should the Web app even know that the mic button should be changed to the muted-state, and that the onclick handler should request-unmute?

**Comment by guidou** [comment]
> @guidou and I seem to agree on focusing on the following items (sorted by priority):

We agree on the items, but IMO the highest priority is to add a mute reason in some form.
Without a mute reason, `requestUnmute` is largely useless as the application won't be able to know when to call it.

**Comment by guidou** [comment]
> I'd support those two MuteReasons, but would avoid exposing more cross-origin correlatable information than that. I don't mind re-firing mute when reason changes.

Only two MuteReasons is not enough to solve the problem since there are more than two reasons for which the track would be muted, with several of them are actionable in different ways. For example, a UA-level mute is actionable via the new `requestUnmute()` API,  while an OS-level mute might be actionable via the same API on some platforms, but only via an informational message on others. Also, it is possible that there are multiple concurrent reasons, with different combinations being actionable (or not) in different ways.

The argument about correlatable information is valid only if we actually solve the problem, so we should solve it in the most privacy-preserving way possible. But we must provide a solution that actually works. If we don't solve the problem (and two mute reasons for sure don't solve the problem) we might as well remove the mute attribute which be more private and just as useful as the attribute in its current form.

**Comment by youennf** [comment]
Exposing mute reasons may expose private information that the user may not want to expose, for instance the fact that the user will take a phone call in the middle of a video conference meeting.
I think it is ok to expose that information when the user is trying to unmute. I do not think it is ok to expose that information to any web page that has capture permission/did capture in the past.

I understand the desire to provide detailed mute reasons. I see it more as a UI optimisation than a blocker for mute management by web pages, given this information would be provided, just later in the game when user actually asks to unmute. At this point, I think it is fine to give precise information.

Hence why I would tend to not start with this but start with the unmute API.
The errors we would expose in the unmute API would help us with unmute reasons, should we want to do this.

> A. Unmuted B. Muted (actionable) C. Muted (unactionable)

I understand this characterisation and agree this is an improvement from a privacy perspective. We could probably implement further mitigations (say expose updated actionable state for visible web pages only for instance).
But again, the web application could go to state B when the mute event is fired and go to state C when the web application asks to unmute and unmuting fails, with a more precise hint of what the actual issue is.

> 2\. How many documents a UA ultimately enforces mute or unmute upon seems implementation-defined.

It would be good if we could agree on a wider scope than track (at least source, maybe device).
If we cannot, so be it. Maybe we could agree on guidelines.

> Without a mute reason, `requestUnmute` is largely useless as the application won't be able to know when to call it.

@guidou, I do not really understand this, can you clarify why a web application won't be able to know when to call?
I would think the web application would call this API whenever the user clicks on the resume capture button.
Do you agree on the assumption that unmuting would be tied to a user gesture?

**Comment by eladalon1983** [comment]
> I understand the desire to provide detailed mute reasons. I see it more as a UI optimisation than a blocker for mute management by web pages, given this information would be provided, just later in the game when user actually asks to unmute.

I still don't understand how the app and user can work to requestUnmute() if the app doesn't even know the upstream is muted. Could you please explain?

**Comment by guidou** [comment]
> Exposing mute reasons may expose private information that the user may not want to expose, for instance the fact that the user will take a phone call in the middle of a video conference meeting. I think it is ok to expose that information when the user is trying to unmute. I do not think it is ok to expose that information to any web page that has capture permission/did capture in the past.

I agree with that. Note that we're talking about pages with a live (but muted) track (not pages that captured in the past but are not currently capturing). I'm fine with restricting the API to that case.  

> I understand the desire to provide detailed mute reasons. I see it more as a UI optimisation than a blocker for mute management by web pages, given this information would be provided, just later in the game when user actually asks to unmute. At this point, I think it is fine to give precise information.
>
I'm fine with restricting more precise information at `requestUnmute()`, after a user gesture.
 
> Hence why I would tend to not start with this but start with the unmute API. The errors we would expose in the unmute API would help us with unmute reasons, should we want to do this.
> 
The problem is that the current `muted` attribute is not good enough for the application to decide when to call `requestUnmute()`. In that sense, having  the two mute reasons @jan-ivar proposed will actually work, since more detailed information will be provided at `requestUnmute` time.  

> > A. Unmuted B. Muted (actionable) C. Muted (unactionable)
> 
> I understand this characterisation and agree this is an improvement from a privacy perspective. We could probably implement further mitigations (say expose updated actionable state for visible web pages only for instance). But again, the web application could go to state B when the mute event is fired and go to state C when the web application asks to unmute and unmuting fails, with a more precise hint of what the actual issue is.
> 

My initial thought is that if the application updates its UI as if `requestUnmute()` was going to be useful,  and then it turns out it's not useful at all, it will be a really bad user experience. But I'm fine with the "actionable"/"not actionable" proposal (or a similar one, such as a new boolean attribute) + more detailed info on `requestUmute()`. 


> > Without a mute reason, `requestUnmute` is largely useless as the application won't be able to know when to call it.
> 
> @guidou, I do not really understand this, can you clarify why a web application won't be able to know when to call? I would think the web application would call this API whenever the user clicks on the resume capture button. Do you agree on the assumption that unmuting would be tied to a user gesture?

I agree on the assumption that unmuting would be tied to a user gesture. I strongly disagree with the assumption that the current `muted` attribute provides the application with enough information of when/how to offer a UI for that user gesture.  We can fix it with a new boolean attribute (I might even call it `muted` if the name wasn't already taken), or with the "actionable"/"not actionable" preliminary mute reason, with extra details provided upon user-initiated unmute.

**Comment by eladalon1983** [comment]
> My initial thought is that if the application updates its UI as if requestUnmute() was going to be useful, and then it turns out it's not useful at all, it will be a really bad user experience. But I'm fine with the "actionable"/"not actionable" proposal (or a similar one, such as a new boolean attribute) + more detailed info on requestUmute().

We should consider that some apps might never wish to call `requestUnmute()`, which might involve a user-prompt by the UA or OS. They might instead prefer to offer a hint - "fix this in the OS" or "fix this in the browser". Having information ahead of calling `requestUnmute()` would help here. The concern around user-privacy, which would motivate us to not expose information ahead of calling `requestUnmute()`, does not strike me as credible; the user trusts the application to listen to their microphone(!) - knowing whether the user muted in the OS or UA is next to no information in comparison. (The app can surmise some mute anyway, given the silence.)

**Comment by jan-ivar** [comment]
> @guidou, I do not really understand this, can you clarify why a web application won't be able to know when to call?

@youennf because apps today don't flip to <img width="16" alt="image" src="https://github.com/w3c/mediacapture-extensions/assets/3136226/2140cb03-78fc-4db8-8a5e-be7e22c107ea"> in response to `mute`. Users would be unable to flip it back, which is bad.

Apps likely want an assurance this was an (actionable) "user" mute before flipping to <img width="16" alt="image" src="https://github.com/w3c/mediacapture-extensions/assets/3136226/2140cb03-78fc-4db8-8a5e-be7e22c107ea">, so that clicking on it is likely to work. Rejection from `unmute()` is too late to determine this.

Apps likely don't want unactionable (non-user) `mute` to affect their mute toggles at all (like today). The goal is to align browser toggles with content toggles.

It's really a separate `userMuted` state in my mind. But Safari already ships a pause UX (which I'd expect to fire _actionable_ mute), so it seems more web compatible to add reason to `mute` than new `usermute` events at this point.

**Comment by jan-ivar** [comment]
If we solve this, then another source of actionable mute/unmute could be the lock screen on phones, if we fix https://github.com/w3c/mediasession/issues/279#issuecomment-1138926629.

**Comment by eladalon1983** [comment]
@jan-ivar, thoughts about the solution [here](https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-1805921604), under `1. Control`, which I intend to present on the coming interim? Note that `"-choice"` vs. `"-issue"` should be a good proxy for what's actionable vs. unactionable.

**Comment by eladalon1983** [comment]
@jan-ivar, thoughts about the solution [here](https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-1805921604), under `1. Listen`, which I intend to present on the coming interim? Note that `"-choice"` vs. `"-issue"` should be a good proxy for what's actionable vs. unactionable.

**Comment by youennf** [comment]
> Rejection from `unmute()` is too late to determine this.

Apps will anyway need to deal with rejection cases, there might be races, there might be user decision changes...
I don't really see much benefit in exposing a boolean about the likelihood of unmute success.

> Apps likely don't want unactionable (non-user) `mute` to affect their mute toggles at all (like today). The goal is to align browser toggles with in-content toggles.

Safari UX is currently showing track muted/unmuted state, whether muting is caused by the user or the OS.
If the web application wants to align with Safari UX, it does not need anything more than the muted boolean value.

From a user perspective, the most important information to convey is whether capture is on or off (off meaning that there is a guarantee It stays off without a user action). I am not sure to understand why the web application would want to present a capture-on icon while the UA UX and OS UX are telling the user that capture is off.

I'd like to understand what UA UX are envisioned for Chrome and Firefox in that space.

**Comment by youennf** [comment]
> A Web application cannot spec-compliantly assume the mic/camera is muted and show relevant app-level UX based on `mute` as it is specified today.

I might buy this definition. Note that this would be a mere boolean in that case (defined as something like UA thinks showing a muted state for the web page makes sense and probably does it on its own UI), not an enumeration.

> Therefore, the `mute` value is **not** a good indication for the Web application that the user muted an upstream entity like the browser or operating system.

Whether the OS muted or the user decided to mute does not change the fact that the UA should present UI to the user that capture on the web page is muted.

I'd like to understand the cases you have in mind where capture for a web page would be muted but UA would show a live capture icon for this web page. AFAIK, there are no such cases in Safari right now, `muted` is all that is needed currently for web pages to update their own UI.

One case I know of is image capture but this seems very specific and we could instead decide to change the image capture spec. Could you list additional cases?

**Comment by eladalon1983** [comment]
> > A Web application cannot spec-compliantly assume the mic/camera is muted and show relevant app-level UX based on `mute` as it is specified today.
>
> I might buy this definition.

Great to hear. So let's discuss how to extend the spec. I proposed an enum. I still think it's a workable solution - I can discuss separately why I think the privacy properties of this proposal are good. But assuming this is not an ideal proposal, what is? Could you please clarify your suggestion of a boolean?

> I'd like to understand the cases you have in mind where capture for a web page would be muted but UA would show a live capture icon for this web page.

Let's examine things not through the lens of macOS and Safari, but through the lens of a hypothetical spec-compliant UA running on some unknown OS. Both the UA and OS offer some form of mic/camera-muting capabilities. Distinctly.

Imagine the user is on Meet/Teams/Zoom and has given mic+camera access. The user has NOT muted anything in the UA. Now let's consider what happens when the user mutes the mic through the OS.

Should the UA change anything it shows the user about mic-access for the app? I believe not, or else the user unmuting the OS-level control would yield a result that could confuse the user - immediately giving the app mic-access again. But the **video-conferencing app** would want to show the mute-icon as changed, because the app speaks to the user of what it **does** with the audio, whereas the UA communicates to the user what **permissions** the app currently has. And the app also wants to allow the user to press this icon, at which point it might call `requestUnmute()`, which would either go directly to the UA, or be relayed from the UA to a similar API in the OS, depending on which mute state(s) apply at that time (the UA knows, even if it does not reveal this info to the Web app).

> One case I know of is image capture but this seems very specific and we could instead decide to change the image capture spec. Could you list additional cases?

I don't understand this question.

**Comment by bradisbell** [comment]
> The mute value can reflect a temporary absence of media for any reason, including mere silence.

For what it's worth, as a developer this is unexpected.  Mute, to me, implies that there is something muting the track, and not simply that silent samples were detected.

I think that is the key to this whole discussion... For something to be muted, something must have muted it.  If something is doing the muting, then that something must be controllable.  The end users expect to find these controls in several locations, including the web application, OS controls, media controls on keyboards, and buttons on headsets.  In the ideal case, the decision to mute or unmute can be made from any of those control surfaces.  Someone should be able to mute on their Bluetooth microphone, and unmute that microphone in the application.

Is there agreement on this principle?

**Comment by guidou** [comment]
> > A Web application cannot spec-compliantly assume the mic/camera is muted and show relevant app-level UX based on `mute` as it is specified today.
> 
> I might buy this definition. Note that this would be a mere boolean in that case (defined as something like UA thinks showing a muted state for the web page makes sense and probably does it on its own UI), not an enumeration.
> 
> > Therefore, the `mute` value is **not** a good indication for the Web application that the user muted an upstream entity like the browser or operating system.
> 
> Whether the OS muted or the user decided to mute does not change the fact that the UA should present UI to the user that capture on the web page is muted.
> 
> I'd like to understand the cases you have in mind where capture for a web page would be muted but UA would show a live capture icon for this web page. AFAIK, there are no such cases in Safari right now, `muted` is all that is needed currently for web pages to update their own UI.
> 

The confusion here is about the current `muted` attribute. If I understand you correctly, you are willing to buy the definition that the `muted` attribute as currently specified does not represent a value that the application can use to update the UI to indicate that the capture is muted.  If so, then we are all in agreement here that we need a new attribute specifically for microphones and cameras.

The new attribute should be defined such that the application can use it to update its UI as you suggest.
IIUC, your might accept a new boolean attribute for this. @jan-ivar proposed a similar one (a 2-value enum with 'actionable', and 'not-actionable' values) and @eladalon1983 is proposing an enum with more detailed values (or a set of such enum values).

Also, IIUC, you are OK with providing more detailed values upon calling `requestUnmute()`, which should be gated by a user gesture indicating intent to unmute.

Perhaps we should start discussing the potential values for the more detailed information and then decide how to make them available. I for sure would like to make a distinction between muting via UA controls, OS settings, and HW (e.g., a dedicated mute button). `requestUnmute` won't be able to unmute all of them, but being able to tell the difference allows the app to provide the user with guidance. I'm OK with the concept of gating values that are more privacy-sensitive behind a user gesture, but I'd like more details about those values and why they require more privacy restrictions than the ones already required for a live track.

**Comment by jan-ivar** [comment]
I like boolean, e.g. an abstract `"actionable"` / `"unactionable"` enum. But nothing more.

But I just realized I think we need it on the `unmute` event as well. So maybe `"user"` / `"temporal"` are better names.

Let me explain:

Examples of temporal mute/unmute might be [bug 1598374](https://bugzil.la/1598374) or where we [dropped the ball on transient activation in gUM](https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-955131893):
> UA could choose to mute for the user if it detects a page accessing camera or microphone on pageload or without interaction.

It's an applications decision how to react to `mute` and `unmute` and how to combine that with the app's own `enabled` state and cam/mic toggles, so knowing whether the `"user"` caused the mute (or unmute!) or whether it's `"temporal"` seems useful.

E.g. many video conferencing lobbies default to the user's mute preferences from last time they joined — so they're not dropped into that corporate-wide meeting with camera and mic unmuted for the world — This is a case where a `"user"` `unmute` maybe should override that preference (and flip the `enabled` toggle), whereas a `"temporal"` `unmute` (concluding an earlier `"temporal"` `mute`) should probably not.

**Comment by jan-ivar** [comment]
The more I think about it I wonder if a separate `track.userMuted` state with `usermute` and `userunmute` events might be better for web compat, to not break apps oblivious to `muteReason`.

Or, maybe we could repurpose the [togglecamera](https://w3c.github.io/mediasession/#dom-mediasessionaction-togglecamera) and [togglemicrophone](https://w3c.github.io/mediasession/#dom-mediasessionaction-togglemicrophone) media session API for this? Isn't their purpose already to toggle the toggles?

E.g. (mic only for brevity)
- User pauses mic in browser: fire `mute` on mic tracks followed by `togglemicrophone` on media session. 
- User unpauses mic in browser: fire `unmute` on mic tracks followed by `togglemicrophone` on media session.

**Comment by youennf** [comment]
I agree with @bradisbell about that the common case (aka user muted capture somewhere and wants a coherent view as well as be able to unmute easily). We can wait to fix the other mute reasons to a later time.

> Perhaps we should start discussing the potential values for the more detailed information and then decide how to make them available.

I agree we should dig into this.
I'd like to start working on `requestMute` and `requestUnmute` model, list the error cases, and learn from those.
That should give us a good foundation to see what additional information we can and should provide at mute time.

> But I just realized I think we need it on the `unmute` event as well. So maybe `"user"` / `"temporal"` are better names?

In terms of API shape, it would probably be better on `MediaStreamTrack` than on the event.
Could you explain how web application would react to `user` vs. `temporal`?

> Apps likely want an assurance this was an (actionable) "user" mute before flipping to <img alt="image" width="16" src="https://user-images.githubusercontent.com/3136226/283266603-2140cb03-78fc-4db8-8a5e-be7e22c107ea.png">, so that clicking on it is likely to work.

This assumes that UA knows whether this is actionable or not at the time of mute.
When mute is triggered by the OS, UA might only know whether this is actionable at the time of calling `requestMute`, not at the time it fires the `mute` event. I prefer web applications to cope by default on this uncertainty.

> * The app's UX speaks to the user of what it **does** with the audio, and whether it has access to audio.
> * The UA's UX communicates to the user what **permissions** the app currently has.

I don't know any UA (or OS) that shows such UX in that way. If these two UX do not represent the same thing but user is confusing them, UA should fix its UX.
In general, I think we should base API design on what UA and OS do, not on what UA and OS might do in the future.

**Comment by alvestrand** [comment]
Is there an app use case for requestMute() that is not covered by track.enabled = false?
I've seen that it's been mentioned on the thread (first time April 20), but searching the thread for "requestMute" unearthed nothing about its imagined use case.

**Comment by youennf** [comment]
> Is there an app use case for requestMute() that is not covered by track.enabled = false?

The main concern is how much web compatible is muting when track.enabled = false.
Web apps currently do not expect that setting `track.enabled=false` will trigger being muted.
Some web apps will set `track.enabled=true` and will not know that they need to call `requestUnmute`.

There might be web applications that want to set `enabled=false` but do not want to be muted, for instance to keep the temporary capture permission that active capture provides.

Third, it is a convenience method which allows to synchronously mute capture by just one call (instead of getting to all tracks that are potentially living in different contexts like workers).

**Comment by eladalon1983** [comment]
> @eladalon1983 is proposing an enum with more detailed values (or a set of such enum values).

Not necessarily just an enum. Another possible approach is to have a few fields for different purposes, such as:
* The mute-source (HW/OS/UA)
* Error vs. choice
* Potential amenability to `requestUnmute()` (see "potential amenability" below)

A `MuteDetails` class could collect these. Some of them might only be exposed if some conditions hold, like a successful call to `requestUnmute()`, or any future thing we introduce.

Concretely:

```webidl
enum MuteSource {"unspecified", "user-agent", "operating-system", "hardware"};

interface MuteReason {
  readonly MuteSource source;
  readonly boolean potentiallyActionable;
  // Possible extension: readonly boolean isError;
};

partial interace MediaStreamTrack {
  sequence<MuteReason> getMutReasons();
};
```

And the spec will say that `mute` is re-fired whenever `sequence<MuteReason>` changes and is non-empty, and `unmute` whenever it changes from non-empty to empty.

> [...] to not break apps oblivious to muteReason.

Apps that assume all mute events are fired in response to an explicit user action are non-compliant (see [earlier quote](https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-1816086419) of spec). Both @jan-ivar and @youennf have in the past argued for the principle that we don't need to worry about breaking non-compliant apps. So we can safely reuse the generic `mute` event and expose `MuteReason`s directly on the track.

> This assumes that UA knows whether this is actionable or not at the time of mute.

Which it often does. Examples:
1. OS-level muting in ChromeOS.
2. UA-level muting in any browser that provides them, such as Safari.

> When mute is triggered by the OS, UA might only know whether this is actionable at the time of calling requestMute, not at the time it fires the mute event.

I propose that one of the `MuteReason` fields should be **potential amenability**. That is, anything where `requestUnmute()` might help with, given the currently available information. (Reasons why it might not work could include heuristics employed by the UA to avoid abuse.)

> I don't know any UA (or OS) that shows such UX in that way. If these two UX do not represent the same thing but user is confusing them, UA should fix its UX.

I believe all UAs work like that. When Chrome shows that Meet is listening to your mic, Chrome cannot tell the user if the application is transmitting the audio remotely, processing it locally, or doing neither, allowing it to go straight to `/dev/null`. Chrome only tells the user - this app has access to your mic right now.

> In general, I think we should base API design on what UA and OS do, not on what UA and OS might do in the future.

Could you please clarify what this statement refers to?
Note that ChromeOS already shows mic/camera controls, and extending such controls to Chromium on arbitrary platforms is something that is frequently discussed.

**Comment by alvestrand** [comment]
> > Is there an app use case for requestMute() that is not covered by track.enabled = false?
> 
> The main concern is how much web compatible is muting when track.enabled = false. Web apps currently do not expect that setting `track.enabled=false` will trigger being muted. Some web apps will set `track.enabled=true` and will not know that they need to call `requestUnmute`.

I was trying to read the spec to figure out if a track will fire onmute or set muted=true when enabled=false.
I couldn't find any language that linked the two; I may have missed it.

**Comment by youennf** [comment]
@guidou , you mentionned muted can be fired in Chrome for various reasons. Can you list them?

**Comment by youennf** [comment]
> I propose that one of the `MuteReason` fields should be **potential amenability**.

So far, for Safari, all mute reasons would be considered actionable (provided I understand what actionable actually means). The only case I know of is https://www.w3.org/TR/image-capture/#dom-imagecapture-takephoto and I question the design here.

I'd like to understand in which existing cases Chrome or Firefox are firing mute events that would not be classified as actionable. That will help understanding/defining this notion and validate whether useful to define/expose or not.
actionability or **potential amenability** is a fuzzy definition, I would prefer we avoid it if we can and instead converge on what mute means for capture tracks.

> I was trying to read the spec to figure out if a track will fire onmute or set muted=true when enabled=false.
> I couldn't find any language that linked the two; I may have missed it.

This was mentioned in https://github.com/w3c/mediacapture-main/issues/642 to handle the case of setting `track.enabled = true` to restart capture (potentially executed by a page in the background, potentially after a long time without capture).

**Comment by youennf** [comment]
Thanks @guidou, this is really helpful info.

> This maps well to the spec text

The spec allows it. I wonder though whether this model is actually helping web developers. For instance, is it better to have a black video or a frozen video when the track is sent via WebRTC?
The stats API that @henbos is working on could be more appropriate for web developers.

FWIW, in Safari, if we do not receive video frames/audio frames after a given amount of time, we fail the capture.
We assume that something is wrong and that the web application be better restarting capture, maybe with a different device. Some web applications do this, not all of them sadly.

These browser differences are making developer's life difficult. I wonder whether this space is now mature enough that we could get browsers to share a more consistent model around muted and capture suspension/failure.
@jan-ivar, how is Firefox using muted these days for capture tracks? Is Firefox sometimes failing capture?

If we cannot converge on the model, I think it is ok to disambiguate things for the web app.
We could go with a boolean value, I do not think differentiating between OS and UA as mute source is providing much benefit.

> there are other non-user-mute reasons

I like this wording and definition better than the previous `actionable` or `amenable` proposed definitions. Basically a capture track was muted following a user decision. `unspecified` and `user-choice` make some sense to me.

@jan-ivar also mentioned MediaSession `togglecamera` and `togglemicrophone`, which could be either done in lieu or in addition to the boolean. This might be a good signal since it is closer to what user does and what OS/UA show to user.
Similarly, `requestToggleCamera`/`requestToggleMicrophone` could have the benefit of making it easier to have consistent UXs between web apps and UA/IOS.

**Comment by guidou** [comment]
> Thanks @guidou, this is really helpful info.
> 
> > This maps well to the spec text
> 
> The spec allows it. I wonder though whether this model is actually helping web developers. 

It's certainly not useful to express muting caused by user actions such as toggling a mute control somewhere (HW/OS).
I don't know if anyone uses it for other purposes, but I wouldn't be surprised if someone does.

> For instance, is it better to have a black video or a frozen video when the track is sent via WebRTC? The stats API that @henbos is working on could be more appropriate for web developers.
> 
> FWIW, in Safari, if we do not receive video frames/audio frames after a given amount of time, we fail the capture. We assume that something is wrong and that the web application be better restarting capture, maybe with a different device. Some web applications do this, not all of them sadly.
> 
> These browser differences are making developer's life difficult. I wonder whether this space is now mature enough that we could get browsers to share a more consistent model around muted and capture suspension/failure. @jan-ivar, how is Firefox using muted these days for capture tracks? Is Firefox sometimes failing capture?
> 
> If we cannot converge on the model, I think it is ok to disambiguate things for the web app. We could go with a boolean value, 

Having a boolean that can be used by the application to know that there is a mute due to user action (or that can be undone via a user action) would be a significant improvement over the current definition. 

> I do not think differentiating between OS and UA as mute source is providing much benefit.

Knowing what specific action the user can take to unmute would allow the application to provide guidance to the user. This would be very beneficial for users. It's common for users to forget how they muted, and often they do it accidentally by unintentionally clicking some toggle somewhere (HW or OS). It's very common to see reports of cameras or microphones "not working" when in fact they had been unintentionally muted by the user. 

> > there are other non-user-mute reasons
> 
> I like this wording and definition better than the previous `actionable` or `amenable` proposed definitions. Basically a capture track was muted following a user decision. `unspecified` and `user-choice` make some sense to me.

I think we can start with that. It would be a significant improvement over the current situation.
We can discuss extending reasons as a follow up by carefully weighing actual privacy risks against the benefit to the user. as part of the `requestToggleX()` or `requestUnmute()` discussion. 

> 
> @jan-ivar also mentioned MediaSession `togglecamera` and `togglemicrophone`, which could be either done in lieu or in addition to the boolean. This might be a good signal since it is closer to what user does and what OS/UA show to user. 
> Similarly, `requestToggleCamera`/`requestToggleMicrophone` could have the benefit of making it easier to have consistent UXs between web apps and UA/IOS.

Can you elaborate on how this would work? Will we need to use the media session API to find out if a track was user-muted? or just that we would reuse those events and fire them on a track? Something else?

**Comment by youennf** [comment]
> Knowing what specific action the user can take to unmute would allow the application to provide guidance to the user.

As said before, I am not against providing this information once user takes action to actually unmute and UA has issues unmuting.

> Can you elaborate on how this would work?

Basically, when user mutes capture with OS/UA UX, UA would fire the mute event on capture tracks and would execute the `togglemicrophone` action callback.
The web page would update its UI within this action callback, either based on capture tracks muted state or on media session additional info. Ditto when user unmutes via OS/UA UX.

**Comment by guidou** [comment]
> As said before, I am not against providing this information once user takes action to actually unmute and UA has issues unmuting.
> 

I don't necessarily agree that some level of detail before user action would be detrimental to privacy, but if we expose the `user-choice` / `unspecified` boolean in some form prior to user action and leave extra details after user action, that would be a significant improvement that we can use to move the discussion forward.

> > Can you elaborate on how this would work?
> 
> Basically, when user mutes capture with OS/UA UX, UA would fire the mute event on capture tracks and would execute the `togglemicrophone` action callback. The web page would update its UI within this action callback, either based on capture tracks muted state or on media session additional info. Ditto when user unmutes via OS/UA UX.

How would this be used in systems with multiple microphones or cameras. with some of them muted and some of them unmuted?  Where would the `user-choice`/`unspecified` value be exposed?

**Comment by eladalon1983** [comment]
As soon as the `mute` event occurs, a Web app should be able to read all of the relevant state. The proposal using `togglemicrophone` does not appear to me to satisfy this requirement. But if we extend `MediaStreamTrack` to expose `getMuteReasons()`, the new state will become immediately exposed at the time `mute` were fired.

**Comment by youennf** [comment]
I agree that the media session action handler should not need to go to MediaStreamTrack to do its processing.
It seems it is missing something that would make it functional, something like:
```
partial MediaSessionActionDetails dictionary {
  bool muting;
}
```
In that case, it seems better to actually design the action handler to execute first, and the mute events to fire second.
This seems consistent with how the spec is designed in general.

Also, maybe we should deprecate `setMicrophoneState` and `setCameraState`.

> How would this be used in systems with multiple microphones or cameras.

I am not sure this is needed, but the `toggle` action scope could be placed in `MediaSessionActionDetails` should the need arise. The scope would be the device, which is the lowest level we should probably go.

**Comment by eladalon1983** [comment]
> I agree that the media session action handler should not need to go to MediaStreamTrack to do its processing. It seems it is missing something that would make it functional, something like:
> 
> ```
> partial MediaSessionActionDetails dictionary {
>   bool muting;
> }
> ```
> 
> In that case, it seems better to actually design the action handler to execute first, and the mute events to fire second. This seems consistent with how the spec is designed in general.

* The interaction between the callback and the `mute` handler is complex and error-prone when multiple mute/unmute actions happen in short succession. A reasonable event listener to `mute` should expect to just be able to read the most recent state without much worry.
* It is unclear how to handle multiple peripherals.
* What about non-mic, non-camera tracks like **screen-sharing**? Note that CrOS already shows UX that allows stopping those; in the future, we might allow pausing (muting).

I think the following proposal is better:

```webidl
interface MuteReason {
  readonly boolean upstream;
};

partial interace MediaStreamTrack {
  sequence<MuteReason> getMuteReasons();
};
```

This is simple, it solves the problem, it is immediately available when the `mute` event fires, and it's fully extensible in the future. For example, we **could** in the future extend MuteReason as:
```webidl
enum MuteSource {"unspecified", "user-agent", "operating-system", "hardware"};

interface MuteReason {
  readonly boolean upstream;
  readonly MuteSource source;
};
```

**Comment by guidou** [comment]
> I agree that the media session action handler should not need to go to MediaStreamTrack to do its processing. It seems it is missing something that would make it functional, something like:
> 
> ```
> partial MediaSessionActionDetails dictionary {
>   bool muting;
> }
> ```
> 
> > How would this be used in systems with multiple microphones or cameras.
> 
> I am not sure this is needed, but the `toggle` action scope could be placed in `MediaSessionActionDetails` should the need arise. The scope would be the device, which is the lowest level we should probably go.

It is essential to know which devices are muted which ones aren't. Multiple cameras and/or microphones is a very common case. The `user-choice`/`unspecified`  values (or whaterver name/form we choose) exposed as mute reasons on MediaStreamTrack look a lot simpler to me and are a straightforward complement to the muted attribute.

Media session looks like it is currently a poor fit that needs a lot of changes to a different spec to support our use case.
The only argument for it is that it has events called `togglemicrophone` and `togglecamera`, which do not add any significant value over what we have in MST, since it already has `mute`  and `unmute` events we can use. The real issue is the new state.

With media session:
* Is there a way to know the initial state if there aren't any toggle events? This is essential to set the correct state in the app UI.
* Is there a way to know the state per device? It would be useless to know that some device is muted if the app cannot tell it is the device currently used by the user.

**Comment by eladalon1983** [comment]
During the editors' meeting, Youenn suggested extending `togglemicrophone` to receive the the mute-state, and possibly make other extensions to address other issues. In that case, the answers to Guido's questions would be "yes":

> Is there a way to know the initial state if there aren't any toggle events?
> Is there a way to know the state per device?

 I think it would **still be unhelpful** to go down that rabbit hole. The show-stoppers are:
1. Microphone and camera are not the only things that can be muted; screen-share is also a concern. Adding `screensharetoggle` and other `xtoggle` would not scale. We don't need a separate API surface for each thing that can be muted.
2. Reasonable Web applications should be able to listen to the `mute` event, read the new state and take action then. This requires an API surface that's updated in conjunction with that event - the Media Session handlers don't fulfil that requirement.

**Comment by eladalon1983** [comment]
I've published a skeleton of a PR in https://github.com/w3c/mediacapture-main/pull/979 - PTAL. If you think `togglemicrophone` is a preferable approach, it would be helpful to see a PR to that effect so we could contrast the two.

**Comment by alvestrand** [comment]
When discussing muting, we should also reflect on the (long) discussion on VideoTrackGenerator.mute - https://github.com/w3c/mediacapture-transform/issues/81

**Comment by youennf** [comment]
I see benefits in MediaSession approach. It is an API used for updating UI/visible application states which is exactly what we are talking about here. It also seems easier to do these updates in a single callback, compared to requiring the web app to potentially coalesce itself multiple mute events.

There are things that should be improved in MediaSession, independently on whether we expose muted reason or not.
I filed some MediaSession issues for that purpose. It makes sense to coordinate with the Media WG on this particular issue.

With regards to the definition of mute reasons, it makes sense to me to piggy-back on MediaSession.
In that sense, it seems ok to me to expose to JS that `MediaStreamTrack` mute/unmute events are triggered by a MediaSession toggle action.

**Comment by eladalon1983** [comment]
To help the discussion culminate in a decision, comparing PRs would be helpful. I have produced a PR for the approach I suggested. @youennf, could you produce a PR for yours?

**Comment by jan-ivar** [comment]
> Here is a media session based proposal:
> 
> For the simple use case (one camera, one microphone), nothing is needed, just use the existing mediaSession API

I like this proposal. I don't see a need to add more information since this seems to be exactly what the mediaSession API was built for (whether the toggles are in a desktop browser UX or on a phone lock screen seems irrelevant). 

Initial state seems solved by firing the mediaSession events early, e.g. on pageload.

This issue is _"Solve user agent camera/microphone double-mute"_, putting other sources out of scope.

Multiple devices also seems out of scope since none of the global UA toggles so far (Safari or Firefox) work per-device AFAIK. They're page or browser global toggles, extending controls present in video conference pages today into the browser, imbuing them with ease of access and some privacy assurance that the webpage cannot hear them, solving the simple use cases of users not being heard, or worrying they can be heard (by participants or webpage). I.e. they affect all devices that page has.

I think Chrome's mute behavior is a bug. I've filed https://github.com/w3c/mediacapture-main/issues/982 to clarify the spec, so let's discuss that there.

I think we should standardize requesting unmute.
I don't think we should standardize requesting mute.
PRs ahead of decisions should not be required.

Too much in this thread.

**Comment by eladalon1983** [comment]
> I don't think we should standardize requesting mute.

Was this suggested at some point?

> PRs ahead of decisions should not be required.

PRs reveal the complexity that otherwise hides behind such phrases as "we could just..."

**Comment by jan-ivar** [comment]
> > I don't think we should standardize requesting mute.
> 
> Was this suggested at some point?

Yes in https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-1244530905.

> We need to solve all use cases that arise in practice, not just the ones indicated in the first message of this thread.

This issue has 70 comments. Triaging discussion out to other (new or existing) issues such as https://github.com/w3c/mediacapture-main/issues/982 or https://github.com/w3c/mediasession/issues/279 seems worthwhile to me, or I don't see how we're going to reach any kind of consensus on all these feature requests. "Mute reason" probably deserves its own issue as well (there were 14 comments here when it was introduced to this conversation in https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-1805921604). It seems largely orthogonal to the OP proposal of letting apps unmute.

> Browser toggles are just one use case that needs to be handled. OS toggles (which can be per device, as in ChromeOS and maybe other OSes) need to be handled too. Hardware toggles need to be considered as well. 

These are all User Agent toggles IMHO, the details of which W3C specs tend to leave to the User Agent, focusing instead on the surface between web app and UA. I think that's the level of abstraction we need to be at.

**Comment by eladalon1983** [comment]
> > > I don't think we should standardize requesting mute.
> > 
> > 
> > Was this suggested at some point?
> 
> Yes in [#39 (comment)](https://github.com/w3c/mediacapture-extensions/issues/39#issuecomment-1244530905).

Thanks for clarifying.
I share your opinion (@jan-ivar) about this proposal.

> Mute reason" probably deserves its own issue [...] It seems largely orthogonal to the OP proposal of letting apps unmute.

Not completely orthogonal, because `requestUnmute()` requires some knowledge of the mute-reason, or else an app would be soliciting a useless user gesture from the user, to their disappointment and frustration.

> These are all User Agent toggles IMHO, the details of which W3C specs tend to leave to the User Agent, focusing instead on the surface between web app and UA. I think that's the level of abstraction we need to be at.

As a representative of one open source browser who has filed bugs and looked into the code of another open source browser, I hope you'll find [this comment](https://github.com/w3c/mediacapture-main/issues/982#issuecomment-1850806625) compelling. It discusses the value transparency brings to the entire ecosystem.

**Comment by jan-ivar** [comment]
Instead of the OP proposal of a `await track.unmute()`, we might already have an API in https://github.com/w3c/mediasession/issues/279#issuecomment-1846023701:
```js
navigator.mediaSession.setMicrophoneActive(false); 
```
E.g. an app calling this with [user attention](https://html.spec.whatwg.org/#fully-active-descendant-of-a-top-level-traversable-with-user-attention) and [transient activation](https://html.spec.whatwg.org/#transient-activation), may be enough of a signal to the UA to unmute tracks it has muted in this document, either raising a toast message about it after the fact, or a prompt ahead of it.

The remaining problem is how the app would learn whether unmuting was successful or not. E.g. might this suffice?
```js
navigator.mediaSession.setMicrophoneActive(false); 
const [unmuted] = await Promise.all([
  new Promise(r => track.onunmute),
  new Promise(r => setTimeout(r, 0))
]);
```

**Comment by youennf** [comment]
setMicrophoneActive looks good to me if we can validate its actual meaning with the media wg.
This API can be extended (return a promise, additional parameters) to progressively cover more of what has been discussed in this thread.

**Comment by jan-ivar** [comment]
> Not completely orthogonal, because requestUnmute() requires some knowledge of the mute-reason, or else an app would be soliciting a useless user gesture from the user, to their disappointment and frustration.

Hiding an unmute control seems a small dent in the disappointment and frustration of being unable to unmute. IOW a secondary problem to the first.

**Comment by eladalon1983** [comment]
> either raising a toast message or a prompt ahead of it.

As I have mentioned multiple times before - the user agent has no idea what "shiny button text" means to the user, or what the user believed he was approving when they conferred transient activation on the page. Only the prompt-based approach is viable.

> Hiding an unmute control seems a small dent in the disappointment and frustration of being unable to unmute.

It does not look at all "small" to me. In fact, I am shocked that after months of debating whether an API should be sync or async, which would have no user-visible effect, you label this issue "small." What is the methodology you employ to classify the gravity of issues?

**Comment by eladalon1983** [comment]
I repeat - there is nothing "small" about a user clicking a button and it disappearing without having an effect. It looks like a bug and it would nudge users towards abandoning the Web app in favor of a native-app competitor. Web developers care much more about their users' perception of the app's reliability, than they do about the inconvenience of [adding async to a method](https://github.com/w3c/mediacapture-region/issues/17). Let's focus our attention where it matters!

**Comment by eladalon1983** [comment]
![image](https://github.com/w3c/mediacapture-extensions/assets/22117736/ac155c5e-95bd-479c-9cf1-871fedd83fb5)

Thank you for this engagement, Jan-Ivar. I am looking forward to hear **why** you disagree.

Orthogonally, I'll be proposing that the rules of conduct in the WG be amended to discourage the use of the thumbs-down emoji without elaboration. Noting disagreement without elaborating on the reasons serves no productive purpose.

**Comment by jan-ivar** [comment]
Closing this as the double-mute problem was instead solved in https://github.com/w3c/mediasession/pull/312.

Here's an [example](https://jsfiddle.net/jib1/psm60wbn/) of how a website can synchronize application mute state with that of the browser.
